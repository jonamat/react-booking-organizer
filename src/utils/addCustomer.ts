import firebase from 'firebase/app';

import { NewCustomer, isNewCustomer, Customer } from '../types';
import { CUSTOMERS_COLLECTION } from '../config';

const REQUIRED_FIELDS = ['name'];

/**
 * Validate user inputs and send them to database
 * @param formData Raw data from form
 * @param birthday Birthday Date instance
 */
const addCustomer = async (formData: FormData, birthday: Date | null): Promise<Customer> => {
    if (REQUIRED_FIELDS.some((fieldName) => !formData.get(fieldName))) throw new Error('Compila i campi obbligatori');

    const newCustomer: any = {
        name: formData.get('name') || undefined,
        surname: formData.get('surname') || undefined,
        additionalIdentifier: formData.get('additionalIdentifier') || undefined,
        phone: formData.get('phone') || undefined,
        birthday: birthday || undefined,
        notes: formData.get('notes') || undefined,
    };

    if (!isNewCustomer(newCustomer)) throw new Error('Errore interno. Contatta il supporto tecnico');

    // Delete empty values
    (Object.keys(newCustomer) as Array<keyof NewCustomer>).forEach((key) => {
        if (newCustomer[key] === undefined) delete newCustomer[key];
    });

    try {
        return {
            id: (await firebase.firestore().collection(CUSTOMERS_COLLECTION).add(newCustomer)).id,
            ...newCustomer,
        };
    } catch (error) {
        console.error(error);
        throw new Error('Connessione al database fallita');
    }
};

export default addCustomer;
