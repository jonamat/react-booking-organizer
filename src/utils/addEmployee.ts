import firebase from 'firebase/app';

import { NewEmployee, Employee, isNewEmployee, WeekWorkingShifts, Holiday } from '../types';
import { EMPLOYEES_COLLECTION } from '../config';

const REQUIRED_FIELDS = ['name'];

/**
 * Validate user inputs and send them to database
 * @param formData Raw data from form
 * @param weekWorkingShifts Valid WeekWorkingShift object
 * @param holidays Valid Holiday array
 */
const addEmployee = async (
    formData: FormData,
    weekWorkingShifts?: WeekWorkingShifts | null,
    holidays?: Array<Holiday> | null,
): Promise<Employee> => {
    if (REQUIRED_FIELDS.some((fieldName) => !formData.get(fieldName))) throw new Error('Compila i campi obbligatori');

    const newEmployee: any = {
        name: formData.get('name') || undefined,
        weekWorkingShifts,
        holidays,
    };

    if (!isNewEmployee(newEmployee)) throw new Error('Errore interno. Contatta il supporto tecnico');

    (Object.keys(newEmployee) as Array<keyof NewEmployee>).forEach((key) => {
        if (newEmployee[key] === undefined) delete newEmployee[key];
    });

    try {
        return {
            id: (await firebase.firestore().collection(EMPLOYEES_COLLECTION).add(newEmployee)).id,
            ...newEmployee,
        };
    } catch (error) {
        console.error(error);
        throw new Error('Connessione al database fallita');
    }
};

export default addEmployee;
