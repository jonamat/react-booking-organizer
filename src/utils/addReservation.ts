import { Employee, Service, Customer, isNewReservation, NewReservation, Reservation } from '../types';
import firebase from 'firebase/app';
import { MaterialUiPickersDate } from '@material-ui/pickers/typings/date';

import { PROGRESSIVE_MINS_WITHIN_HOUR, MIN_STEP, TIME_DATE_FORMAT, RESERVATIONS_COLLECTION } from '../config';
import { isEmployeeBusy, isEmployeeOnDuty } from '.';

const REQUIRED_FIELDS = ['serviceId'];

interface Database {
    employees: Array<Employee>;
    services: Array<Service>;
    reservations: Array<Reservation>;
}

/**
 * Validate user inputs and send them to database
 * @param formData Raw data from form
 * @param customer Customer requesting the reservation
 * @param selectedDate Date of the reservation
 * @param validationData Redux data from store.database for validation
 * @param forceAssignment Force assignment even if employee in not on duty
 */
const addReservation = async (
    formData: FormData,
    customer: Customer | null,
    selectedDate: MaterialUiPickersDate,
    { employees, services, reservations }: Database,
    forceAssignment?: boolean,
): Promise<Reservation> => {
    if (REQUIRED_FIELDS.some((fieldName) => !formData.get(fieldName)) || !customer)
        throw new Error('Compila i campi obbligatori');

    if (!selectedDate || (typeof selectedDate === 'number' && isNaN(selectedDate)))
        throw new Error('Seleziona una data valida');

    if (!PROGRESSIVE_MINS_WITHIN_HOUR.includes(selectedDate.getMinutes()))
        throw new Error(`L'orario indicano non rispetta lo scaglionamento orario dell'agenda.
        Utilizza intervalli di ${MIN_STEP} minuti.`);

    const newReservation: any = {
        date: selectedDate,
        customerId: customer.id,
        serviceId: formData.get('serviceId') || undefined,
        employeeId: formData.get('employeeId') || undefined,
        notes: formData.get('notes') || undefined,
    };

    if (!isNewReservation(newReservation)) throw new Error('Errore interno. Contatta il supporto tecnico');

    const humanReadableDate = newReservation.date.toLocaleDateString('it-IT', TIME_DATE_FORMAT);

    const employee = employees.find((employee) => employee.id === newReservation.employeeId);
    if (!employee) throw new Error('Errore interno - Dipendente non trovato');

    if (isEmployeeBusy(newReservation, services, reservations))
        throw new Error(`${employee?.name} è già impegnato/a con un altro appuntamento ${humanReadableDate}.`);

    if (!forceAssignment && !isEmployeeOnDuty(newReservation.date, employee)) throw new Error('EMPLOYEE_NOT_ON_DUTY');

    (Object.keys(newReservation) as Array<keyof NewReservation>).forEach((key) => {
        if (newReservation[key] === undefined) delete newReservation[key];
    });

    try {
        return {
            id: (await firebase.firestore().collection(RESERVATIONS_COLLECTION).add(newReservation)).id,
            ...newReservation,
        };
    } catch (error) {
        console.error(error);
        throw new Error('Connessione al database fallita');
    }
};

export default addReservation;
