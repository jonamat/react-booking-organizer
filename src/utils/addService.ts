import firebase from 'firebase/app';

import { NewService, isNewService, Service } from '../types';
import { PROGRESSIVE_MINS_WITHIN_MAX_SERVICE, SERVICES_COLLECTION } from '../config';

const REQUIRED_FIELDS = ['name', 'averageDuration'];

/**
 * Validate user inputs and send them to database
 * @param formData Raw data from form
 */
const addService = async (formData: FormData): Promise<Service> => {
    if (REQUIRED_FIELDS.some((fieldName) => !formData.get(fieldName))) throw new Error('Compila i campi obbligatori');

    const newService: any = {
        name: formData.get('name') || undefined,
        averageDuration: parseInt(formData.get('averageDuration') as string) || undefined,
    };

    if (!isNewService(newService) || !PROGRESSIVE_MINS_WITHIN_MAX_SERVICE.includes(newService.averageDuration))
        throw new Error('Errore interno. Contatta il supporto tecnico');

    (Object.keys(newService) as Array<keyof NewService>).forEach((key) => {
        if (newService[key] === undefined) delete newService[key];
    });

    try {
        return {
            id: (await firebase.firestore().collection(SERVICES_COLLECTION).add(newService)).id,
            ...newService,
        };
    } catch (error) {
        console.error(error);
        throw new Error('Connessione al database fallita');
    }
};

export default addService;
