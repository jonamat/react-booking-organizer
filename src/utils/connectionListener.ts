import { isOffline } from '../redux';

/**
 * Trigger a state change on connection changes
 * @param dispatch Redux dispatch function
 */
const connectionListener = (dispatch) => {
    // First check
    dispatch(isOffline(!navigator.onLine));

    window.addEventListener('online', () => dispatch(isOffline(false)));
    window.addEventListener('offline', () => dispatch(isOffline(true)));
};

export default connectionListener;
