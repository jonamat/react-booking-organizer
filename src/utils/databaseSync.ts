import firebase from 'firebase/app';
import { bindActionCreators } from 'redux';
import { updateServices, updateEmployees, updateCustomers, updateReservations } from '../redux';
import { SERVICES_COLLECTION, EMPLOYEES_COLLECTION, CUSTOMERS_COLLECTION, RESERVATIONS_COLLECTION } from '../config';

/**
 * Synchronize database changes with redux store
 * @param dispatch Redux dispatch function
 * @returns An array of unsubscribe functions
 */
const databaseSync = (dispatch): Array<() => void> => {
    const unsubscribeObservers = Array<() => void>(4);
    const actions = bindActionCreators(
        { updateServices, updateEmployees, updateCustomers, updateReservations },
        dispatch,
    );

    unsubscribeObservers[0] = firebase
        .firestore()
        .collection(SERVICES_COLLECTION)
        .onSnapshot((collectionSnapshot) => {
            actions.updateServices(collectionSnapshot);
        });

    unsubscribeObservers[1] = firebase
        .firestore()
        .collection(EMPLOYEES_COLLECTION)
        .onSnapshot((collectionSnapshot) => {
            actions.updateEmployees(collectionSnapshot);
        });

    unsubscribeObservers[2] = firebase
        .firestore()
        .collection(CUSTOMERS_COLLECTION)
        .onSnapshot((collectionSnapshot) => {
            actions.updateCustomers(collectionSnapshot);
        });

    unsubscribeObservers[3] = firebase
        .firestore()
        .collection(RESERVATIONS_COLLECTION)
        .onSnapshot((collectionSnapshot) => {
            actions.updateReservations(collectionSnapshot);
        });

    return unsubscribeObservers;
};

export default databaseSync;
