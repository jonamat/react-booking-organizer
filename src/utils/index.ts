// Utils
export { default as isEmployeeBusy } from './isEmployeeBusy';
export { default as isEmployeeOnDuty } from './isEmployeeOnDuty';
export { default as generateClosestValidDate } from './generateClosestValidDate';
export { default as getReservationsByDate } from './getReservationsByDate';
export { default as getReservationStatus } from './getReservationStatus';

export { default as formatMinutes } from './formatMinutes';
export { default as hourMinuteToReadableString } from './hourMinuteToReadableString';
export { default as translateWeekDays } from './translateWeekDays';

export { default as databaseSync } from './databaseSync';
export { default as connectionListener } from './connectionListener';

// Auth system
export { default as login } from './login';
export { default as logout } from './logout';

// Delete database elements
export { default as deleteCustomer } from './deleteCustomer';
export { default as deleteReservation } from './deleteReservation';
export { default as deleteEmployee } from './deleteEmployee';
export { default as deleteService } from './deleteService';

// Add database elements
export { default as addCustomer } from './addCustomer';
export { default as addEmployee } from './addEmployee';
export { default as addReservation } from './addReservation';
export { default as addService } from './addService';

// Edit database elements
export { default as editCustomer } from './editCustomer';
export { default as editEmployee } from './editEmployee';
export { default as editReservation } from './editReservation';
export { default as editService } from './editService';
