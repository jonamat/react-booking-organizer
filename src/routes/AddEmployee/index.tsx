import React, { FC, useRef, FormEvent, useState } from 'react';
import { RouteComponentProps } from 'react-router';
import { Typography, TextField, Button, Divider, TextFieldProps, CircularProgress } from '@material-ui/core';
import { toast } from 'react-toastify';

import { WeekWorkingShifts, Holiday } from '../../types';
import { addEmployee } from '../../utils';
import HolidaysField from '../../components/HolidaysField';
import ShiftsField from '../../components/ShiftsField';
import useStyles from './style';

const AddEmployee: FC<RouteComponentProps> = ({ history }) => {
    const classes = useStyles();
    const [loading, setLoading] = useState<boolean>(false);
    const [weekWorkingShifts, setWeekWorkingShifts] = useState<WeekWorkingShifts | null>(null);
    const [holidays, setHolidays] = useState<Array<Holiday> | null>(null);
    const inputForm = useRef<HTMLFormElement>(null);
    const fieldProps: TextFieldProps = {
        className: classes.field,
        variant: 'outlined',
        autoComplete: 'off',
    };

    const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
        event.preventDefault();
        if (!inputForm.current) return;

        // Disable user interactions
        setLoading(true);

        addEmployee(new FormData(inputForm.current), weekWorkingShifts, holidays)
            .then((employee) => {
                toast.success('Aggiunto al personale: ' + employee.name);
                history.push('/employees');
            })
            .catch((error) => {
                toast.error(error.message);
                setLoading(false);
            });
    };

    return (
        <div className={classes.container} data-testid="add-employee">
            <Typography variant="h6">Aggiungi dipendente</Typography>
            <Divider className={classes.divider} />
            <form className={classes.form} ref={inputForm} onSubmit={handleSubmit}>
                <TextField {...fieldProps} inputProps={{ 'data-testid': 'name' }} name="name" label="Nome" required />

                <ShiftsField
                    classes={{ root: classes.field }}
                    value={weekWorkingShifts}
                    onSelect={setWeekWorkingShifts}
                />

                <HolidaysField classes={{ root: classes.field }} value={holidays} onSelect={setHolidays} />

                <div className={classes.submitButton}>
                    <Button
                        type="submit"
                        color="primary"
                        variant="contained"
                        disabled={loading}
                        startIcon={loading ? <CircularProgress size={20} /> : null}
                        data-testid="submit-button"
                    >
                        {loading ? 'Caricamento' : 'Aggiungi'}
                    </Button>
                    <Button variant="contained" color="primary" disabled={loading} onClick={() => history.goBack()}>
                        Annulla
                    </Button>
                </div>
            </form>
        </div>
    );
};

export default React.memo(
    AddEmployee,
    (prevProps, nextProps) => prevProps.location.search === nextProps.location.search,
);
