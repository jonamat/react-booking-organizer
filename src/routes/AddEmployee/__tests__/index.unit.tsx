/**
 * Unit test suite
 *
 * Target: AddEmployee component
 * Tests:
 * - Responses handling
 * - Redirection
 * - User interaction interruption on loading
 * - Middleware function call
 * - Visual feedbacks
 * - Snapshot matching
 * - Functionality
 *
 * For data validation, API call and errors handling see {@link '../../../utils/addEmployee'}
 */

// Test resources
import React from 'react';
import renderer from 'react-test-renderer';
import { StylesProvider, createGenerateClassName } from '@material-ui/core';
import { render, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

// Test configuration
import { addEmployee } from '../../../utils';

// Mocked modules
import { toast } from 'react-toastify';

// Component to test
import AddEmployee from '..';

/* -------------------------------------------------------------------------- */
/*                                    Mocks                                   */
/* -------------------------------------------------------------------------- */

jest.mock('../../../utils', () => ({
    addEmployee: jest.fn(() => Promise.resolve({ id: 'fakeID', name: 'Carl' })),
}));
jest.mock('../../../components/HolidaysField', () => () => null);
jest.mock('../../../components/ShiftsField', () => () => null);
const spyTSuccess = jest.spyOn(toast, 'success');
const spyTError = jest.spyOn(toast, 'error');
const history = {
    push: jest.fn(),
};

/* -------------------------------------------------------------------------- */
/*                              Bridge operations                             */
/* -------------------------------------------------------------------------- */

beforeEach(() => {
    jest.clearAllMocks();
});

/* -------------------------------------------------------------------------- */
/*                               Render wrapper                               */
/* -------------------------------------------------------------------------- */

/**
 * Render the component with the custom configuration
 * @param props Component props
 */
const renderComponent = (props?) => {
    return render(<AddEmployee history={history} {...props} />);
};

/* -------------------------------------------------------------------------- */
/*                                    Tests                                   */
/* -------------------------------------------------------------------------- */

describe('AddEmployee component', () => {
    describe('Upload system', () => {
        it('calls the db middleware function with correct user inputs', () => {
            const { getByTestId } = renderComponent();

            fireEvent.input(getByTestId('name'), { target: { value: 'name' } });
            fireEvent.click(getByTestId('submit-button'));

            expect(addEmployee as jest.Mock).toBeCalledTimes(1);

            expect(Array.from((addEmployee as jest.Mock).mock.calls[0][0].entries())).toEqual([['name', 'name']]);
            expect((addEmployee as jest.Mock).mock.calls[0][1]).toBeNull();
            expect((addEmployee as jest.Mock).mock.calls[0][2]).toBeNull();
        });
    });

    describe('User experience', () => {
        it('redirects user on success', async () => {
            const { getByTestId } = renderComponent();

            fireEvent.input(getByTestId('name'), { target: { value: 'name' } });
            fireEvent.click(getByTestId('submit-button'));

            await waitFor(() => {
                expect(history.push).toBeCalledTimes(1);
                expect(history.push).toBeCalledWith('/employees');
            });
        });

        it('calls toast on success', async () => {
            const { getByTestId } = renderComponent();

            fireEvent.input(getByTestId('name'), { target: { value: 'name' } });
            fireEvent.click(getByTestId('submit-button'));

            await waitFor(() => {
                expect(spyTSuccess).toBeCalledTimes(1);
            });
        });

        it('calls toast on error', async () => {
            (addEmployee as jest.Mock).mockImplementation(() => Promise.reject('some error'));

            const { getByTestId } = renderComponent();

            fireEvent.input(getByTestId('name'), { target: { value: 'name' } });
            fireEvent.click(getByTestId('submit-button'));

            await waitFor(() => {
                expect(spyTError).toBeCalledTimes(1);
            });
        });

        it('disables submit button on loading', async () => {
            (addEmployee as jest.Mock).mockImplementation(() => Promise.reject('some error'));

            const { getByTestId } = renderComponent();

            fireEvent.input(getByTestId('name'), { target: { value: 'name' } });
            expect(getByTestId('submit-button').hasAttribute('disabled')).toBeFalsy();
            fireEvent.click(getByTestId('submit-button'));
            expect(getByTestId('submit-button').hasAttribute('disabled')).toBeTruthy();

            await waitFor(() => {
                expect(getByTestId('submit-button').hasAttribute('disabled')).toBeFalsy();
            });
        });
    });

    it('matches snapshot', () => {
        const snapshot = renderer
            .create(
                <StylesProvider generateClassName={createGenerateClassName({ disableGlobal: true })}>
                    {/* @ts-ignore */}
                    <AddEmployee />
                </StylesProvider>,
            )
            .toJSON();
        expect(snapshot).toMatchSnapshot();
    });
});
