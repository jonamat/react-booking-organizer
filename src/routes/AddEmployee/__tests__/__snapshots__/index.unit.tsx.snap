// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddEmployee component matches snapshot 1`] = `
<div
  className="makeStyles-container-1"
  data-testid="add-employee"
>
  <h6
    className="MuiTypography-root-6 MuiTypography-h6-16"
  >
    Aggiungi dipendente
  </h6>
  <hr
    className="MuiDivider-root-36 makeStyles-divider-2"
  />
  <form
    className="makeStyles-form-3"
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root-44 MuiTextField-root-43 makeStyles-field-4"
    >
      <label
        className="MuiFormLabel-root-60 MuiInputLabel-root-48 MuiInputLabel-formControl-54 MuiInputLabel-animated-57 MuiInputLabel-outlined-59 MuiFormLabel-required-66 MuiInputLabel-required-52"
        data-shrink={false}
      >
        Nome
        <span
          aria-hidden={true}
          className="MuiFormLabel-asterisk-67 MuiInputLabel-asterisk-53"
        >
           
          *
        </span>
      </label>
      <div
        className="MuiInputBase-root-83 MuiOutlinedInput-root-68 MuiInputBase-formControl-84"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          autoFocus={false}
          className="MuiInputBase-input-94 MuiOutlinedInput-input-78"
          data-testid="name"
          disabled={false}
          name="name"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
        />
        <fieldset
          aria-hidden={true}
          className="PrivateNotchedOutline-root-101 MuiOutlinedInput-notchedOutline-77"
        >
          <legend
            className="PrivateNotchedOutline-legendLabelled-103"
          >
            <span>
              Nome
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      className="makeStyles-submitButton-5"
    >
      <button
        className="MuiButtonBase-root-134 MuiButton-root-105 MuiButton-contained-113 MuiButton-containedPrimary-114"
        data-testid="submit-button"
        disabled={false}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label-106"
        >
          Aggiungi
        </span>
      </button>
      <button
        className="MuiButtonBase-root-134 MuiButton-root-105 MuiButton-contained-113 MuiButton-containedPrimary-114"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label-106"
        >
          Annulla
        </span>
      </button>
    </div>
  </form>
</div>
`;
